### DISCUSSION TITLE
Greenter en Laravel

### DISCUSSION THREAD

**Discussion by jeanomar13** [thegreenter/greenter#153] [General]
Estoy intentando instalar y usar Greenter en mi proyecto desarrollado en Laravel, pero tengo dudas que me gustaria que me las resolvieran:
- Instale Greenter a través del comando  composer require greenter/lite
- Luego cree una clase con el código siguiente

namespace Jolc\Repository;
use Illuminate\Support\Facades\Storage;
use Greenter\Ws\Services\SunatEndpoints;
use Greenter\See;

class Sunat{

  public function getSee() {

    $see = new See();
    $see->setService(SunatEndpoints::FE_BETA);
    $certificate = Storage::disk('storage')->get('Sunat/cert.pem');
    if ($certificate === false) {
      throw new Exception('No se pudo cargar el certificado');
    }
    $see->setCertificate($certificate);
    $see->setClaveSOL('20000000001', 'MODDATOS', 'moddatos');
    return $see;
  }
}

El codigo de mi controlador es:

namespace App\Http\Controllers;
use Illuminate\Http\Request;
use App\Models\Repository\Sunat;

use Greenter\Model\Client\Client;
use Greenter\Model\Company\Company;
use Greenter\Model\Company\Address;
use Greenter\Model\Sale\Invoice;
use Greenter\Model\Sale\SaleDetail;
use Greenter\Model\Sale\Legend;
use Greenter\See;
class MyController extends Controller
{
    protected $arrays, $sunat;

    /**
     * Create a new controller instance.
     *
     * @return void
     */
    public function __construct(Sunat $sunat)
    {
        $this->sunat = $sunat;
    }
    
   /**
     * Show the form for editing the specified resource.
     *
     * @param  int  $id
     * @return \Illuminate\Http\Response
     */
    public function myFunction(Request $request)
    {
        **return $this->sunat->getSee();**

y cuando llamo a la clase desde mi controlador en myFunction() tengo el siguiente error:
TypeError
Argument 1 passed to Symfony\Component\HttpFoundation\Response::setContent() must be of the type string or null, object given, called in C:\xampp\htdocs\cursosweb\vendor\laravel\framework\src\Illuminate\Http\Response.php on line 65

**Comment by giansalex** [comment]
Hola, `$see` es un objeto, para generar la factura y enviarla a sunat, puedes usar este [ejemplo](https://github.com/thegreenter/demo/blob/master/examples/factura.php#L76).

**Reply by jeanomar13** [reply]
Hola muchas gracias por contestar mi pregunta, estuve revisando el ejemplo que mencionas y aun me queda esta duda:
para el ejemplo se instancia $util = Util::getInstance(); para poder usar $see = $util->getSee(SunatEndpoints::FE_BETA);
y obtengo este error en mi controlador Class 'App\Http\Controllers\Util' not found
he probado tu proyecto y funciona pero en cuanto trato de llevar el código a laravel tengo ese error, agradecería mucho puedas ayudarme con eso

**Reply by MrJmpl3** [reply]
Que tengas un error de eso, me hace suponer que eres nuevo en PHP, recomiendo aprender bien PHP puro antes de usar algun Framework como Laravel. Ya que asi es el orden correcto de la enseñanza y practica.

Y la solucion a tu error, es que no estas importando el Namespace, igualmente el demo es una demostracion, lo que debe hacer es interpretar el codigo del demo y aplicarlo en Laravel

**Reply by jeanomar13** [reply]
Con el debido respeto mrjmpl, 3 lo que tu mencionas de importar el namespace es obvio, pero cual es el namespace para Util a eso va mi pregunta. Si estas aquí para criticar el nivel de programación de cada uno, lo siento pero no quise ofender el nivel que puedas tener

**Reply by MrJmpl3** [reply]
Si te ofendi, lo lamento, aunque era un consejo no lo tomes a mal. No tiene nada de malo ser nuevo en algo, todos lo fuimos y lo seguimos siendo en muchas cosas.

Te explico:
- Vemos que el ejemplo, carga el autoload: https://github.com/thegreenter/demo/blob/master/examples/factura.php#L11
- En composer esta la organizacion de los namespace y carpetas: https://github.com/thegreenter/demo/blob/master/composer.json#L16
- Vemos que el namespace vacio esta organizado a la carpeta src, ademas vemos que Util no tiene imports, por ello suponemos que tiene el namespace vacio y por ende debe estar en la carpeta src: https://github.com/thegreenter/demo/blob/master/src/Util.php
- A ti te sale el error del Util, porque? Por los namespace, Laravel pre-configura los namespace porque al inicio del todo, ya importa el autoloader y seleccionar que carpeta va con cada namespace.

Solucion:
- Verifica que hace la funcion que necesitamos de Util: https://github.com/thegreenter/demo/blob/master/src/Util.php#L44
- Aplicar aparte las funciones que necesitas de Util: https://github.com/thegreenter/greenter/discussions/153#discussioncomment-216266

Como te dice giansalex, `Util` solo se encuentra en demo, Y vuelvo a repetir lo mismo de mi comentario anterior, porque demo es una opcion para probar, por lo siguiente: demo no es una opcion para ir, copiar y usar, sino para leer, interpretar y modificar y probar.

**Reply by jeanomar13** [reply]
Gracias

**Comment by giansalex** [comment]
Hola @jeanomar13, no necesitas usar `Util` (esto solo se encuentra en demo), ya que es lo mismo que hace tu método `Sunat::getSee()`, podrías usarlo directamente.
```php
$see = $this->sunat->getSee();
$res = $see->send($invoice);

```

PD: Aquí se encuentra la [clase Util](https://github.com/thegreenter/demo/blob/master/src/Util.php), si necesitas algún método puedes copiarlo a tu proyecto.

**Reply by jeanomar13** [reply]
Muchisimas gracias por el apoyo @giansalex, la verdad que me quedo todo más claro, ya con tu explicación podre avanzar en mi proyecto :+1: .... YA obtuve mi primer mensaje : La Factura numero F001-1, ha sido aceptada

**Reply by tony98ms** [reply]
@jeanomar13  Hola amigo, pudiste hacer la integracion con laravel, me gustaria que me ayudes, recien voy hacer la implementacion, si puedes este es mi numero o dejame el tuyo +593 63227447
